brave/brave-core
# Time-limited sync words

Resolves https://github.com/brave/brave-browser/issues/22242

This PR introduces time-limited sync code words. The browser accept the old sync code till the grace period, currently `1 Jun 2022` - but may be changed. Changes are done for Desktop and Android.

## Submitter Checklist:

- [x] There is a [ticket](https://github.com/brave/brave-browser/issues/22242) for my issue

## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:

I. Old sync words are accepted before sync words sunset date
1. Create chain on old desktop browser, 24 words;
2. Connect new desktop browser to the chain - it should connect;
3. Connect new Android browser to the chain with words - it should connect.
4. Connect new Android browser to the chain with QR - it also should connect.


II. Old sync words are not accepted after sync words sunset date
1. Create chain on old desktop browser, 24 words;
    magic room blind room mimic suspect interest fiber above south end injury quality wonder maid book syrup brush right alter fit weasel buffalo stove
2. Modify current date on the desktop to 2nd of June;
3. Try to connect new desktop browser to the chain - it should say 
`This code was generated by a deprecated version of Brave, please upgrade your other device first`;
4. Modify current date on the desktop to 2nd of June;
5. Try to connect new Android browser to the chain with words - it also should say 
`This code was generated by a deprecated version of Brave, please upgrade your other device first`;

III. Normal flow between new versions
1. Create sync chain on new  desktop version, 25 words.
2. Connect new browser desktop to chain via words - it should succeed;
3. Connect new browser Android to chain via words - it also should succeed;

IV. Case for expired error
1. Create sync chain on new  desktop version, 25 words.
2. Adjust time on 2nd desktop device 2 days forward;
3. Connect new browser desktop to chain via words - it should say 
`This code has expired. Please generate a new one on your other device and try again`;
4. Adjust time on Android device 2 days forward;
5. Connect new browser Android to chain via words - it also should say
`This code has expired. Please generate a new one on your other device and try again`;

V. Case for too far in the future error
1. Create sync chain on new  desktop version, 25 words.
2. Adjust time on 2nd desktop device 2 days backwards;
3. Connect new browser desktop to chain via words - it should say 
`This code is invalid. Please check that the time and timezone are set correctly on your device`;
4. Adjust time on 2nd Android device 2 days backwards;
5. Connect new browser Android to chain via words - it also should also say
`This code is invalid. Please check that the time and timezone are set correctly on your device`.

VI. Case for general error of wrong code
1. Try to enter the sync chain on new dektop with words `AAA BBB CCC` - it should say
`This code is invalid`
2. Try to enter the sync chain on new Android with words `AAA BBB CCC` - it should also say
`Wrong sync code`
